hugr-py/src/hugr/utils.py

```diff
@@ -0,0 +1,53 @@
+from collections.abc import Hashable, ItemsView, MutableMapping
+from dataclasses import dataclass, field
+from typing import Generic, TypeVar
+
+
+L = TypeVar("L", bound=Hashable)
+R = TypeVar("R", bound=Hashable)
+
+
+@dataclass()
+class BiMap(MutableMapping, Generic[L, R]):
+    fwd: dict[L, R] = field(default_factory=dict)
+    bck: dict[R, L] = field(default_factory=dict)
+
+    def __getitem__(self, key: L) -> R:
+        return self.fwd[key]
+
+    def __setitem__(self, key: L, value: R) -> None:
+        self.insert_left(key, value)
+
+    def __delitem__(self, key: L) -> None:
+        self.delete_left(key)
+
+    def __iter__(self):
+        return iter(self.fwd)
+
+    def __len__(self) -> int:
+        return len(self.fwd)
+
+    def items(self) -> ItemsView[L, R]:
+        return self.fwd.items()
+
+    def get_left(self, key: R) -> L | None:
+        return self.bck.get(key)
+
+    def get_right(self, key: L) -> R | None:
+        return self.fwd.get(key)
+
+    def insert_left(self, key: L, value: R) -> None:
```

Comment:
There is no check here that `value` isn't already the value of another key. This could lead to inconsistent bimaps. (Similarly for `insert_right`.)

Reply:
won't it just overwrite that entry (like a dict would) ?

Reply:
It will overwrite the entry in `self.bck` but not the entry in `self.fwd`. So a subsequent call to `get_right(other_key)` will still return `value`.

Reply:
should be fixed now